File: src/java/nz/co/searchwellington/urls/UrlBuilder.java
package nz.co.searchwellington.urls;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.text.SimpleDateFormat;
import java.util.Date;

import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.Newsitem;
import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.SiteInformation;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.model.UrlWordsGenerator;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.model.Website;

public class UrlBuilder {

	private SiteInformation siteInformation;

		
	public UrlBuilder(SiteInformation siteInformation) {		
		this.siteInformation = siteInformation;
	}

	public String getImageUrl(String filename) {
		return siteInformation.getImageRoot() + filename;
	}
	
	public String getStaticUrl(String filename) {
		return siteInformation.getStaticRoot() + filename;
	}
	
	public String getPublishersAutoCompleteUrl() {
		return siteInformation.getUrl() + "/ajax/publishers";
	}
	
	public String getTagsAutoCompleteUrl() {
		return siteInformation.getUrl() + "/ajax/tags";
	}
		
	public String getFeedUrl(Feed feed) {		
		return siteInformation.getUrl() + "/feed/" + feed.getUrlWords();
	}
	
	public String getFeedsInboxUrl() {
		return siteInformation.getUrl() + "/feeds/inbox";
	}
	
	public String getFeedsUrl() {
		return siteInformation.getUrl() + "/feeds";
	}
	
	public String getTagUrl(Tag tag) {
		return siteInformation.getUrl() + "/" + tag.getName();
	}
	
	
	public String getAutoTagUrl(Tag tag) {
		return siteInformation.getUrl() + "/autotag/" + tag.getName();
	}

	public String getTagCombinerUrl(Tag firstTag, Tag secondTag) {
		return siteInformation.getUrl() + "/" + firstTag.getName() + "+" + secondTag.getName();
	}
	
	public String getTagSearchUrl(Tag tag, String keywords) {
		try {
			return siteInformation.getUrl() + "/search?keywords=" + URLEncoder.encode(keywords, "UTF-8") + "&tag="+ URLEncoder.encode(tag.getName(), "UTF-8");
		} catch (UnsupportedEncodingException e) {
			return null;
		}
	}
	
	public String getLocalPageUrl(Newsitem newsitem) {
		return siteInformation.getUrl() + UrlWordsGenerator.markUrlForNewsitem(newsitem);
	}
	
	public String getPublisherUrl(Website publisher) {
		return siteInformation.getUrl() + "/" + publisher.getUrlWords();
	}

	public String getPublisherCombinerUrl(Website publisher, Tag tag) {
		return siteInformation.getUrl() + "/" + publisher.getUrlWords() + "+" + tag.getName();
	}

	public String getTagCommentUrl(Tag tag) {
		return siteInformation.getUrl() + "/" + tag.getName() + "/comment";
	}
	
	public String getTagGeocodedUrl(Tag tag) {
		return siteInformation.getUrl() + "/" + tag.getName() + "/geotagged";
	}

	public String getCommentUrl() {
		return siteInformation.getUrl() + "/comment";
	}

	public String getJustinUrl() {
		return siteInformation.getUrl() + "/justin";
	}

	public String getGeotaggedUrl() {
		return siteInformation.getUrl() + "/geotagged";
	}
	
	
	public String getPublicTaggingUrl(Resource resource) {
		return siteInformation.getUrl() + "/tagging/tag?resource=" + resource.getId();
	}
	
	public String getTaggingUrl(Resource resource) {
		return siteInformation.getUrl() + "/edit/tagging?resource=" + resource.getId();
	}
	
	public String getArchiveUrl() {
		return siteInformation.getUrl() + "/archive";
	}
	
	public String getArchiveLinkUrl(Date date) {
		// TODO if this dateformatter thread safe? Replace with DateFormatter
		SimpleDateFormat df = new SimpleDateFormat();
	    df.applyPattern("yyyy");
	    String yearString = df.format(date.getTime());
	    df.applyPattern("MMM");
	    String monthString = df.format(date.getTime());
	    return siteInformation.getUrl() + "/archive/" + yearString + "/" + monthString.toLowerCase();		
	}

	public String getOpenIDCallbackUrl() {
		return siteInformation.getUrl() + "/openid/callback";
	}
	
	public String getProfileUrl(User user) {
		return siteInformation.getUrl() + "/profiles/" + user.getProfilename();
	}

	public String getWatchlistUrl() {
		return siteInformation.getUrl() + "/watchlist";
	}
	
}

File: src/main/scala/nz/co/searchwellington/urls/UrlBuilder.scala
package nz.co.searchwellington.urls

import nz.co.searchwellington.controllers.models.helpers.ArchiveMonth
import nz.co.searchwellington.filters.attributesetters.PageParameterFilter
import nz.co.searchwellington.model._
import nz.co.searchwellington.model.frontend.{FrontendFeed, FrontendNewsitem, FrontendResource, FrontendWebsite}
import nz.co.searchwellington.model.geo.OsmId
import org.joda.time.Interval
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import uk.co.eelpieconsulting.common.geo.model.Place

import java.time.LocalDate

@Component
class UrlBuilder @Autowired()(siteInformation: SiteInformation, urlWordsGenerator: UrlWordsGenerator) extends ArchiveMonth {

  def fullyQualified(uri: String): String = {
    siteInformation.getUrl + uri
  }

  def getHomeUri: String = "/"

  def getImageUrl(filename: String): String = getStaticUrl(filename)

  def getStaticUrl(filename: String): String = {
    siteInformation.getStaticRoot + filename
  }

  def getPublishersUrl: String = {
    getHomeUri + "publishers"
  }

  def getPublishersAutoCompleteUrl: String = {
    "/ajax/publishers"
  }

  def getTagsAutoCompleteUrl: String = {
    "/ajax/tags"
  }

  def getFeedUrl(feed: FrontendFeed): String = {
    "/feed/" + feed.getUrlWords
  }

  def getFeedUrl(feed: Feed): String = {
    "/feed/" + feed.url_words.getOrElse("")
  }

  def getFeedUrlFromFeedName(feedname: String): String = {
    "/feed/" + urlWordsGenerator.makeUrlWordsFromName(feedname)
  }

  def getFeedsInboxUrl: String = {
    "/feeds/inbox"
  }

  def getFeedsUrl: String = {
    "/feeds"
  }

  def getNewFeedForPublisherUrl(publisher: FrontendWebsite): String = {
    "/new-feed?publisher=" + publisher.urlWords
  }

  def getTagsUrl(): String = "/tags"

  def getTagUrl(tag: Tag): String = {
    "/" + tag.name
  }

  def getTagPageUrl(tag: Tag, page: Long): String = {
    getTagUrl(tag) + "?page=" + page
  }

  def getAutoTagUrl(tag: Tag): String = {
    "/" + tag.getName + "/autotag"
  }

  def getTagCombinerUrl(firstTag: Tag, secondTag: Tag): String = {
    "/" + firstTag.getName + "+" + secondTag.getName
  }

  def getTagCombinerUrl(firstTag: Tag, secondTag: Tag, page: Int): String = {
    getTagCombinerUrl(firstTag, secondTag) + "?page=" + page
  }

  def getLocalPageUrl(resource: FrontendResource): String = {
    resource match {
      case n: FrontendNewsitem =>
        "/newsitem/" + n.id
      case f: FrontendFeed =>
        "/feed/" + f.getUrlWords
      case w: FrontendWebsite =>
        "/" + w.urlWords
      case _ =>
        null
    }
  }

  def getLocalPageUrl(resource: Resource): String = {
    resource match {
      case n: FrontendNewsitem =>
        "/newsitem/" + n.id
      case f: FrontendFeed =>
        "/feed/" + f.getUrlWords
      case w: FrontendWebsite =>
        "/" + w.urlWords
      case _ =>
        ""  // TODO Watchlist items have no local page
    }
  }

  def getPublisherUrl(publisher: Website): String = "/" + publisher.url_words.get // TODO Naked get
  def getPublisherUrl(publisher: FrontendWebsite): String = "/" + publisher.urlWords

  def getPublisherPageUrl(publisher: Website, page: Int): String = {
    getPublisherUrl(publisher) + "?page=" + page
  }

  def getPublisherCombinerUrl(publisher: Website, tag: Tag): String = {
    "/" + publisher.url_words.get + "+" + tag.getName // TODO Naked get - make url words mandatory for publishers
  }

  def getPublisherTagCombinerUrl(publisher: FrontendWebsite, tag: Tag): String = {
    "/" + publisher.urlWords + "+" + tag.getName
  }

  def getTagCommentUrl(tag: Tag): String = {
    "/" + tag.getName + "/comment"
  }

  def getTagGeocodedUrl(tag: Tag): String = {
    "/" + tag.getName + "/geotagged"
  }

  def getCommentUrl: String = {
    "/comment"
  }

  def getAcceptedUrl: String = "/accepted"

  def getAcceptedUrl(date: LocalDate): String = "/accepted?date=" + date.toString

  def getJustinUrl: String = {
    "/justin"
  }

  def getGeotaggedUrl: String = {
    "/geotagged"
  }

  def getArchiveUrl: String = {
    "/archive"
  }

  def getArchiveLinkUrl(link: IntervalLink): String = {
    link match {
      case p: PublisherArchiveLink =>
        getPublisherArchiveLinkUrl(p)
      case t: TagArchiveLink =>
        getTagArchiveLinkUrl(t)
      case _ =>
        "/archive/" + renderYearMonth(link.interval)
    }
  }

  def getIntervalUrl(interval: Interval): String = {
    "/archive/" + renderYearMonth(interval)
  }

  def getPublisherArchiveLinkUrl(link: PublisherArchiveLink): String = {
    "/" + link.publisher.getUrlWords + "/" + renderYearMonth(link.interval)
  }

  def getTagArchiveLinkUrl(link: TagArchiveLink): String = {
     getTagUrl(link.tag) + "/" + renderYearMonth(link.interval)
  }

  def getOpenIDCallbackUrl: String = {
    "/openid/callback"
  }

  def getProfileUrl(user: User): String = {
    getProfileUrlFromProfileName(user.getProfilename)
  }

  def getProfileUrlFromProfileName(username: String): String = {
    "/profiles/" + UrlParameterEncoder.encode(username)
  }

  def getWatchlistUrl: String = {
    "/watchlist"
  }

  def getTwitterCallbackUrl: String = siteInformation.getUrl + "/twitter/callback"

  def getLocationUrlFor(place: Place): String = {
    if (place.getOsmId != null) {
      "/geotagged?osm=" + UrlParameterEncoder.encode(place.getOsmId.getId + "/" + place.getOsmId.getType)
    } else if (place.getLatLong != null) {
      "/geotagged?latitude=" + place.getLatLong.getLatitude + "&longitude=" + place.getLatLong.getLongitude
    } else {
      null
    }
  }

  def getLocationUrlForRadius(place: Place, radius: Double): String = {
    if (place.getOsmId != null) {
      "/geotagged?osm=" + UrlParameterEncoder.encode(place.getOsmId.getId + "/" + place.getOsmId.getType) + "&radius=" + radius
    } else if (place.getLatLong != null) {
      "/geotagged?latitude=" + place.getLatLong.getLatitude + "&longitude=" + place.getLatLong.getLongitude + "&radius=" + radius
    } else {
      null
    }
  }

  def getSearchUrlFor(q: String): String = getSearchUrlFor(q, None, None)

  def getSearchUrlFor(keywords: String, page: Option[Int] = None, tag: Option[Tag] = None, publisher: Option[Website] = None): String = {
    val ps = Seq(
      page.map(p => PageParameterFilter.PAGE_ATTRIBUTE -> p.toString),
      tag.map(t => "tag" -> t.name),
      publisher.map(p => "publisher" -> p.url_words.get)  // TODO naked get
    ).flatten

    io.lemonlabs.uri.Url(path = "/search").addParam("q", keywords).addParams(ps).toRelativeUrl.toString()
  }

  def getTagSearchUrl(tag: Tag, keywords: String): String = {
    getSearchUrlFor(keywords, tag = Some(tag))
  }

  def getPublisherSearchUrl(publisher: Website, keywords: String): String = {
    getSearchUrlFor(keywords, publisher = Some(publisher))
  }

  def getSubmitFeedUrl: String = {
    "/new-feed"
  }

  def getSubmitNewsitemUrl: String = {
    "/new-newsitem"
  }

  def getSubmitWatchlistUrl: String = "/new-watchlist"

  def getSubmitWebsiteUrl: String = {
    "/new-website"
  }

  def getResourceUrl(resource: FrontendResource): String = {
    getLocalPageUrl(resource)
  }

  def getOsmWebsiteUrl(osmId: OsmId): String = {
    "http://www.openstreetmap.org/browse/" + osmId.getType.toLowerCase + "/" + osmId.getId
  }

  def getDiscoveredFeeds: String = getFeedsUrl + "/discovered"

}
